src/java.base/share/classes/jdk/internal/loader/URLClassPath.java

```diff
@@ -439,34 +439,56 @@ private synchronized Loader getLoader(int index) {
                 continue;
             }
             // Otherwise, create a new Loader for the URL.
-            Loader loader;
+            Loader loader = null;
+            final URL[] loaderClassPathURLs;
             try {
                 loader = getLoader(url);
-                // If the loader defines a local class path then add the
-                // URLs as the next URLs to be opened.
-                URL[] urls = loader.getClassPath();
-                if (urls != null) {
-                    push(urls);
-                }
+                // If the loader defines a local class path then construct
+                // and add the URLs as the next URLs to be opened.
+                loaderClassPathURLs = loader.getClassPath();
             } catch (IOException e) {
-                // Silently ignore for now...
+                // log the error and close the unusable loader (if any)
+                if (DEBUG) {
+                    System.err.println("Failed to construct a loader or construct its" +
+                            " local classpath for " + url + ", cause:" + e);
+                }
+                if (loader != null) {
+                    closeLoaderIgnoreEx(loader);
+                }
                 continue;
             } catch (SecurityException se) {
-                // Always silently ignore. The context, if there is one, that
-                // this URLClassPath was given during construction will never
-                // have permission to access the URL.
+                // log the error and close the unusable loader (if any).
+                // The context, if there is one, that this URLClassPath was
+                // given during construction will never have permission to access the URL.
                 if (DEBUG) {
                     System.err.println("Failed to access " + url + ", " + se );
                 }
+                if (loader != null) {
+                    closeLoaderIgnoreEx(loader);
+                }
                 continue;
             }
+            if (loaderClassPathURLs != null) {
+                push(loaderClassPathURLs);
+            }
             // Finally, add the Loader to the search path.
             loaders.add(loader);
             lmap.put(urlNoFragString, loader);
         }
         return loaders.get(index);
     }
 
+    // closes the given loader and ignores any IOException that may occur during close
+    private static void closeLoaderIgnoreEx(final Loader loader) {
```

Comment:
closeQuietly(Loader loader) would work too.

Reply:
I had considered `safeClose(...)` but that didn't sound right. But what you suggest is more precise. I've updated the PR to use this suggestion.